feat(discover2) Add high-level API for getting discover facets

markstory opened the pull request:
Right now we half let users define which tags they want in the facet bars, and then use many HTTP requests and snuba queries to get that information. This new API will allow us to use fewer HTTP requests and more importantly fewer snuba queries to fetch tag facets.

The project facet requires special casing as it is not actually a tag. This API will also allow us to add more clever logic to deciding which results to show in the future as well.

markstory committed:
feat(discover2) Add high-level API for getting discover facets

Right now we half let users define which tags they want in the facet
bars, and then use many HTTP requests and snuba queries to get that
information. This new API will allow us to use fewer HTTP requests and
more importantly fewer snuba queries to fetch tag facets.

The project facet requires special casing as it is not actually a tag.
This API will also allow us to add more clever logic to deciding which
results to show in the future as well.

fpacifici commented on src/sentry/snuba/discover.py:
```diff
@@ -415,3 +417,103 @@ def get_pagination_ids(event, query, params, reference_event=None, referrer=None
+        projects = [
+            FacetResult("project", r["project_id"], r["count"]) for r in project_values["data"]
+        ]
+        results.extend(projects)
```
What is the goal of this section ?

markstory replied:
Part of the facets are to show event distributions across the various projects. This adds the project facet values to the returned results.

fpacifici commented on src/sentry/snuba/discover.py:
```diff
@@ -415,3 +417,103 @@ def get_pagination_ids(event, query, params, reference_event=None, referrer=None
+        dataset=Dataset.Discover,
+        limit=limit,
+        referrer=referrer,
+        turbo=sample,
```
I think we should force turbo over a given number of projects like we do on tagstore 
https://github.com/getsentry/sentry/blob/master/src/sentry/tagstore/snuba/backend.py#L249-L260

markstory replied:
I can do that. :+1:

markstory replied:
I chose 10 projects as the point to enable sampling. We can tweak it as necessary.

markstory committed:
Make sampling automatically enable.

Turn sampling on after 9 projects. I've chosen 9 arbitrarily as
a starting point.

markstory committed:
Use the promoted environment column

Using the tags array for environments will skip environment='' records
which we want to consider.

dashed reviewed (approved):
Looks good to me 👍
